### ISSUE TITLE
NX 2.0 iterator API

### ISSUE THREAD

**Issue by dschult** [networkx/networkx#1246]
As we move to the iterator reporting paradigm I think it is worth spelling out what the reporting API should be. My hope is that we can make it easier to build graph classes that don't rely on dict-of-dict-of-dict data structures. see also #980 

I'll include a list of all methods further down. This first post only discusses reporting methods impacted by the iterator-change and/or d-o-d-o-d data structure.  See #1152 for similar treatement of .copy() 

The big change I am proposing (in addition to returning iterators) is that the keyword "data" can be None, True or anything-else.  None means return no data(as before), True means to return the edge data as a dict (as before), anything-else means to return the edge-attribute-value associated with the edge-attribute-key stored in "data"; essentially G[u][v][data].  Notice that we would need to rule out edge attributes with key "True".

Proposed API for Reporting from Di/Multi/Graph
- nodes() - iterator over nodes
- neighbors(n, data=None) - iterator over neighbors of n. 
  -      If data is True: iterator over (nbr, eattr_dict)
  -      if data not in (None,True): iterator over (nbr, eattr_dict[data]) 
- degree(nbunch=None, weight=None) - iterator over (n, deg(n))
  -     if nbunch is not None: only iterate over nodes in nbunch… single node->single value
  -     if weight is not None: deg(n) is sum of weights on edges instead of number of edges 
- edges(nbunch=None, data=None, keys=None) - iterator of 2-tuples (undirected only once)
  -     if nbunch is not None: iterator of edges from nodes in nbunch (undirected only once)
  -     if data is True: iterator of edge info (n,nbr,eattr_dict)
  -     if data not in (None,True): 3-tuple (n,nbr,eattr_dict[data]
  - MultiGraph
    -     if data/key is True/None: iterator of edge info (n,nbr,keydict)
    -     if data/key is None/True: iterator of 3-tuple edge (n,nbr,key)
    -     if data/key is True/True: iterator of 4-tuple edge (n,nbr,key,eattr_dict)
    -     if data not in (True,None): 4-tuple (n,nbr,key,eattr[data]) 
- get_edge_data(u, v, data=None, key=None) eattr_dict for edge between u,v
  -     if data is not None: eattr[data] for edge between u,v
  -     For MultiGraphs if key is None, return eattr_keydict, otherwise look at edge u,v,key
- adjacency_iter(data=None) - iterator of (n, {nbr:eattr, ...})  This is dict-of-dict... better way?? 
  -     if data is not None: eattr replaced by eattr[data]

**Comment by dschult** [comment]
Here's another issue we have wrestled with in the past that should be decided during our discussion of the 2.0 API changes. 

Should the API differ for Di/Multi/Graphs?  For example, we have in/out_edges for directed but not for undirected. We have edges(keys=True) only for multi.

It may seem strange to have in/out_edges for undirected graphs. But leaving it out means we have lots of places in our codebase where we have to check is_multigraph or is_directed in order to figure out how to call the method. It would be slicker if we unified the APIs so the algorithms could just use a general call signature and the object would figure out what to do. 

For example, if the algorithm uses edge keys, we currently choose either to only support multigraphs, or we add code to check is_multigraph and more code to add keys to each edge of a non-multi. With a unified API we would use Graph.edges(keys=True), expecting a 3-tuple and even a Graph object would oblige with a (u,v,0) 3-tuple.

Similarly, if the algorithm uses in_edges() a unified API would allow that call to a Graph or a DiGraph. The Graph would treat each edge as an in_edge. The algorithm code wouldn't have to worry about which base class is being used.

Are there cases where this wouldn't work? That is, where it is ambiguous what the method would want from a more simple class? I can't come up with one.

The design specification on the Wiki is written as if all keywords are available for all graph classes. It was not my intention to do that, but as I think about it maybe that would be a good idea. It could remove code from many routines that change the call signature depending on di/multi/class.

**Comment by chebee7i** [comment]
I can think of one case where a unified API won't work: [mixed graphs](http://en.wikipedia.org/wiki/Mixed_graph).

But we don't even support mixed graphs right now, and all the existing algorithms would have to be modified to support them anyway. Graphs like these come up in probabilistic graphical models under the name "chain graphs" or "partially directed acyclic graphs" (see Probabilistic Graphical Models by Koller and Friedman).

**Comment by MridulS** [comment]
I opened a new issue for the design specification #1382

**Comment by dschult** [comment]
Reading through this Issue, I think we have either implemented or split off to another issue all the concepts discussed here. So I will close this issue and point people to #1631 #1168 #2467 #1382 #1344  #1336 #1080
